src/org/rust/lang/core/resolve/util/RustResolveUtil.kt

```diff
@@ -10,22 +13,49 @@ public object RustResolveUtil {
         while (current != null) {
             when (current) {
                 is RustResolveScope -> return current
-                else -> current = current.parent
+                else                -> current = current.parent
             }
         }
 
         return null
     }
 
-    fun getGlobalResolveScopeFor(elem: PsiElement): RustResolveScope {
-        var current = getResolveScopeFor(elem)
-        while (true)
-        {
-            val p = current?.let { getResolveScopeFor(it) } ?: break
-            current = p;
+    fun getCrateRootFor(elem: PsiElement): RustResolveScope? {
+        val mod = getSelfModFor(elem) ?: return null
+
+        val superMod = getSuperModFor(mod)
+        return if (superMod == null) mod else getCrateRootFor(superMod)
+    }
+
+    fun getSuperModFor(elem: RustModItem): RustModItem? {
+        val self = getSelfModFor(elem) ?: return null
+        val superInFile = self.containingMod
+        if (superInFile != null) {
+            return superInFile
         }
 
-        return current!!;
+        val file = self.containingFile
+        val dir = self.containingFile?.containingDirectory ?: return null
+        val dirOfParent = if (self.ownsDirectory) dir.parent else dir
+        dirOfParent?.files.orEmpty()
+            .filterIsInstance<RustFileImpl>()
+            .map { it.mod }
+            .filterNotNull()
+            .forEach { mod ->
+                for (declaration in mod.modDecls) {
+                    val childFile = declaration.moduleFile as? ChildModFile.Found ?: continue
+
+                    if (childFile.file == file ) {
+                        return mod
+                    }
+                }
+            }
+
+        return null
+    }
+
+    fun getSelfModFor(elem: PsiElement): RustModItem? {
+        return elem.parentOfType<RustModItem>(strict = false)
```

Comment:
Docs are in progress. 

Maybe we don't need `containtgMod` at all and should use `getSelfModFor` instead. But I'm not sure about this (equally, we may want to provide getSelf/getSuper as extensions). 

I'd suggest to keep things as they are until we have a better understanding of the desirable API.